Improve usability of non-default values in the property inspector

RandomShaper opened the pull request:
Implements https://github.com/godotengine/godot-proposals/issues/4893.

- Provide a visual indication that a (sub)group contains non-default (revertable) values **when it's collapsed**.
- Add a new option to the inspector's tools menu for expanding only (sub)groups containing properties with non-default values.
- ~Add an editor setting to enable/disable such behavior.~ (Removed.)

![non_defaults](https://user-images.githubusercontent.com/11797174/181915938-0e949a36-69ab-4ae2-85a0-9f09ae1e5b7f.gif)

(Updated a million times.)

that-richard commented on doc/classes/EditorProperty.xml:
```diff
@@ -101,6 +101,13 @@
+			<argument index="0" name="property" type="StringName" />
+			<argument index="1" name="can_revert" type="bool" />
+			<description>
+				Emitted when the tbe revertability (i.e., whether it has a non-default value and thus is displayed with a revert icon) of a property has changed.
```
"the tbe revertability"

RandomShaper replied:
Good catch. Fixed.

Calinou commented:
> Add an editor setting to enable/disable such behavior.

Is this really needed? We should only add new editor settings for features that will be disabled by a significant amount of users, and can't live with the feature enabled for a good reason.

RandomShaper commented:
@Calinou, I've done the following:

- Removed the editor setting. If many people wanted it, we can always add it later.
- Added 1.5x spacing between the icon and the group caption and fixed the spacing between the collapse/expand arrow that I had made narrower than formerly by mistake.

groud commented:
I think the feature is ok. I don't have a need for it but it seems like a lot of users need strong visual clues that something is not using the default value.

I think the icon isn't good enough though. IMO, the revert icon is an icon for executing an action, not really something showing a status. I think we should maybe use another symbol, maybe a star or something like that? We could keep the arrow in a corner but I don't think the arrow should be the main element of the icon.

Calinou commented:
> I think the icon isn't good enough though. IMO, the revert icon is an icon for executing an action, not really something showing a status. I think we should maybe use another symbol, maybe a star or something like that?

I guess we could always display a `*` character next to the category name.

akien-mga commented:
> I guess we could display a `*` character next to the category name (perhaps with a leading space or two), so we don't need an icon.

Note that `(*)` is what we use to identify scenes with unsaved changes. Reusing a similar character might be confusing and make users think that they have unsaved changes in that Node/Resource.

RandomShaper commented:
> I think the feature is ok. I don't have a need for it but it seems like a lot of users need strong visual clues that something is not using the default value.
> 
> I think the icon isn't good enough though. IMO, the revert icon is an icon for executing an action, not really something showing a status. I think we should maybe use another symbol, maybe a star or something like that? We could keep the arrow in a corner but I don't think the arrow should be the main element of the icon.

That's why it's not exactly the same icon. It's the same inner shape, which shows it's related to revert, but different at the same time due to the inverted colors and placed somewhere that there are no buttons, which hints further it's not clickable.

Since this is alpha, we can keep it as it is and see if users complain. At this point, discussion is a shot in the dark game.

arkology commented:
Feature is great, but icon looks too "heavy".

Diddykonga commented:
I also agree I am not a fan of the Icon as is, as Arkology mentioned above, it being a full circle makes it look heavy relative to its peers.

I agree with Calinou in using `*` or the original revert icon.

RandomShaper commented:
I see suggestions in different directions, so I've tried to improve my own approach to make everyone maybe a bit happier. Namely, I've made the icon smaller. I think it's hard to confuse it with a clickable button, since this one is very different from all the Godot icon buttons, which are icons floatings on a transparent background. Also, being the same color of the text adds another hints that it's lifeless as the text, exactly as if it was the asterisk some suggest, only fancier.

cassandrus commented:
What if make it greyed out? Expand/Collapse icons are darker than texts & the new icon.
Also, there is a "pin" icon, also greyed out:
![image](https://user-images.githubusercontent.com/97176349/180953595-c03832f2-73d9-4fec-88e4-892a0dd28996.png)
Looks unclickable.

reduz commented:
My concern with this is that I don't understand why a property that would have changed would be folded. As things are:

* You have to unfold to edit, and folding is kept. If it's not it should be filed as a bug.
* If you never edited this object, Godot will unfold all properties that were edited the first time you did an edit.

So, in which cases would you re-fold (or never un-fold) an edited property? I think this is more interesting to research to me before moving forward with a PR like this.

RandomShaper commented:
@cassandrus, the pic icon is indeed as gray as the text, but at least it reacts to hover and looks like other actionable icons (a little image with no background). The fact the collapse/fold arrows are a different color than the text is a hint that they are not static, like the text is. That's precisely a reason to use the text color for the new icon.

@reduz, regarding reasons to re-fold, you may want to stay focused in properties both above and below a group; therefore you'd want to fold that group in the middle. Or maybe one just that you have the habit of collapsing everything you're not actively using at the moment, to get clutter out of the inspector. Some user mentioned this in the proposal and I can at least say that's my experience, too.

Now, when you open back a scene for editing, or you are handed a scene edited by other person whose folding habits you don't know or just doesn't have a strict folding discipline, which is the most likely case, you probably want to explore which properties in some nodes you or the other person touched. This exploration is easier if you can see quickly where there are properties with non-default values.

Or maybe you are checking a just imported scene and want to explore how your stuff from Blender ended up represented in Godot.  (Maybe this is a case of unfolding that should happen automatically?)

arkology commented:
> Or maybe one just that you have the habit of collapsing everything you're not actively using at the moment, to get clutter out of the inspector.

Exactly.

reduz commented:
I would do the following changes:
1) Do not use an icon, use something like (*) or just * at the end of the name.
2) Do not show this when unfolded. I think the main issue is to see it when its folded.

ajreckof commented:
> 2. Do not show this when unfolded. I think the main issue is to see it when its folded.

I strongly disagree with this one just as you can have a folded group with edited properties in it you can have an unfolded group with nothing edited in it for example if you defaulted some values. Making it inconsistent will make it confusing and i don't see a reason to hide it when you unfold it

RandomShaper commented:
I'm starting to see that those that don't like/need the feature are suggesting discreet ways to represent the non-default group status so it annoys them less. But that's like striving for an average solution that won't please anyone much.

Isn't the solution, then, to keep a full-fledged indicator (an icon shown whether the group is folded or not) for those who want it and have an editor setting so the others can just have it completely disabled? The discussion would be then about what the default should be. I'd agree on it being conservatively disabled.

Calinou commented:
> I'd agree on it being conservatively disabled.

New features that are disabled by default have poor discoverability. They basically don't exist to users until they find them out :slightly_smiling_face:

reduz commented:
@RandomShaper Indeed, its a quite invasive feature for those who don't need it (as I mentioned above, many if not most users  never re-fold sections so they don't need this feature), so given that the problem is that you don't know which folded sections contain modified properties the solution should focus exclusively on it.

As such, it only should be an indicator visible when the section is folded. As a suggestion, I would try simply adding an asterisk to the section name when folded, which is common language in Godot to show that something is modified:
![image](https://user-images.githubusercontent.com/6265307/181599913-72b1641e-a24e-482f-89a2-97f8794b7f04.png)

If we see that this is eventually not enough, we can try something else, but so far as its common practice for us lets start with the absolute minimum.

reduz commented:
@ajreckof Sorry but that does not make any sense. For us who don't re-fold sections this has never been a problem and the PR is not aiming to solve this either. The revert buttons are plainly visible on their own when the section is unfolded.

The actual problem has been well described: You don't know if you have modified properties when the section is folded.
This is what the PR needs to focus on solving.

ajreckof commented:
I like the idea of keeping it simple but the asterisk is really not a good idea. While you say that asterisk is indicating that something is modified but for me and i think a lot of others an asterisk is only indicating that something has not yet been saved. 
Moreover if you really wish to keep the indicator only when folded (i still don't see why you do) maybe having something underneath like a small colored line would be better as it would naturally disappear when expending the group

reduz commented:
@ajreckof As I said, I am not sure it matters its not exactly the same. This is the simplest, least invasive way to implement this so lets try first and see how it works out. If for users its not clear, we change it. Otherwise we keep it.

RandomShaper commented:
1. I see the point in hiding the mark when the group is expanded (I don't like it personally, but it could be controlled via an editor setting, say `docks/property_editor/hide_non_default_mark_in_expanded_groups`).
2. I fail to see why the asterisk is superior to the icon. The last design of the icon is quite subtle, doesn't look like a poor man's alternative that a text-based GUI would have to restort to, and won't trigger in users what the asterisk would cause: frustration of not being able to get rid of it to release the cognitive weight the regular use of the asterisk has.

reduz commented:
@RandomShaper I think we should only add a settings if they actually are a problem and a demand for them. For now we don't know if it's a problem. Let people use it and see after a while how they find the feature.

Regarding the icon, in the PR meeting we did today nobody really seemed to like it much or understand it, and many mentioned that by looking at it one could easily think its something you should be able to click and that it does something (like any other icon in the inspector). I think an icon is not the way for this, which is why I suggested trying with an *.

RandomShaper commented:
OK. I've updated the code with the minimalistic approach. I'm keeping a copy of the former version in a separate branch, just in case. I've also updated the images in the PR description.

kleonc commented:
Another possible approach: actually add revert buttons to the (sub)groups. Would use the same icon as existing ones. Would be shown if there's any revertible property in the given (sub)group. Clicking it would revert all such properties.

Sounds easy and intuitive but the problem would be visual placement of such buttons. Already existing revert buttons for properties are aligned nicely and thus they don't feel like a clutter at all. But I don't see how these potential new revert buttons for (sub)groups could be placed so they would look consisent with the already existing ones. And if they won't look consistent with already existing stuff then it would result in the same clutter feel reduz mentioned in the case of that new revert icon indicator placed to the right of the (sub)group name.
So meh unless someone would find a way to make things look natural/consistent.

---

Regarding just an `*` I think it might actually be enough but my concern is it looks mysterious, it's not obvious what it means. Hence it would be nice if it would be at least somehow explained/mentioned in the tooltip for (sub)groups marked with `*`. E.g. something like:
```
* there are revertible properties in this group.
```

reduz commented:
@kleonc Yeah I also agree the asterisk is a bit mysterious but anything is pretty much unless you unfold it.  Its quite challenging to convey this better.

An alternative I am thinking could be something like this:

![image](https://user-images.githubusercontent.com/6265307/181647172-4e13fd49-8fc7-4df3-8962-66b7ab3e7298.png)

This is 100% clear to the user about what is going on.

RandomShaper commented:
Someone suggested another count based approach: https://twitter.com/FilippoArtemis/status/1551537006232100866.

Maybe less explicit than text, but more elegant. I'm happy with whichever more than the asterisk.

reduz commented:
@RandomShaper If you are not happy with the asterisk, I would go with the text based approach I mentioned since that (unlike an icon) is the easiest one to understand.

RandomShaper commented:
I've tagged this as _Needs work_ and I'll implement it and update the PR soon.

RandomShaper committed:
Improve usability of non-default values in the property inspector

- Provide a visual indication that a (sub)group contains non-default (revertable) values when it's collapsed.
- Add a new option to the inspector's tools menu for expanding only (sub)groups containing properties with non-default values.

RandomShaper commented:
@reduz, I've implemented it. Check the PR description to see the result. RTL is also well supported.

I've changed some more code to properly support that subgroup unfolding trigger unfolding of their parent group.

reduz commented:
seems CI did not complete on MacOS for reasons unrelated

akien-mga commented:
Thanks!

dmaz commented:
@RandomShaper 
just a guess but this introduced a bug that duplicates the first inspector panel items if those items are not in a group.
![image](https://user-images.githubusercontent.com/203130/182388198-ec4600bd-5646-4e9a-9c90-dce5519ad6af.png)

YuriSizov commented:
@dmaz That is probably not related to this PR. It's either https://github.com/godotengine/godot/issues/63454, which is fixed, or https://github.com/godotengine/godot/issues/63668, which is going to be fixed.

dmaz commented:
oh sweet!  thanks.  I should of looked into it more.   this pr was just in my head cause it's actually fricken awesome!
